### PR TITLE
SNS/ SQS Lambda instrumentation

### PR DESCRIPTION
**Pull request by Vlaaaaaaad**
Hi,

I am a bit on the fence about this PR hence the _Draft PR_.

A common pattern is to have SNS/SQS -> other services/ Lambda -> SNS/SQS -> Lambda. Having to load the trace data from the message each time a Lambda is invoked is annoying and I'd much rather have the beeline get it directly from there.
This PR adds support for auto-discovery of the `x-honeycomb-trace` data from SNS and SQS Message Attributes.

While working on this though I also kinda wanted to add support for DynamoDB Streams and for actually getting the Honeycomb data from the message itself -- the data should really be in Message Attributes but that is such an annoying call.
An idea came to me: **what if the decorator would be able to take a parameter and take the data from there**?
Say we have `@beeline_wrapper('event/body/honeycomb-data')` or anything like that. Dynamically parsing the object as needed( yes this would get into the body in some cases where the user desires that which would be not efficient but awesome from a UX perspective).
I am unsure about how wise of an idea this is. What do you think about it?

### CODE REVIEW

**Comment by tredman** [comment]
Hey, sorry for letting this sit for days. I am not super familiar with dynamo or some of the other AWS message structures. Can you elaborate on:

> the data should really be in Message Attributes but that is such an annoying call.

Do they all support message attributes? It would be nice to standardize on a message attribute key for trace data and not have to support looking into arbitrary locations in message bodies.

I am curious: how are you adding trace context to SNS/SQS messages now? Manually? What if we could patch the boto lib to inject trace context transparently, then we could rely on the context being in the same place each time. We do something like this for the `requests` lib to inject the X-Honeycomb-Trace header. Boto is more complex, but it could be done. Would that be a better user experience?

**Comment by Vlaaaaaaad** [comment]
No worries. This was/ is not at all urgent -- I would've pinged you if it was. The project I was planning to use this on even got postponed a bit.

>  we could patch the boto lib to inject trace context transparently, then we could rely on the context being in the same place each time. We do something like this for the requests lib to inject the X-Honeycomb-Trace header. Boto is more complex, but it could be done. Would that be a better user experience?

I did not even think of this. I love the idea!

> > the data should really be in Message Attributes but that is such an annoying call.
> 
> Do they all support message attributes? It would be nice to standardize on a message attribute key for trace data and not have to support looking into arbitrary locations in message bodies.

That would've been nice, but of course life is not that simple. Let me expand:

### SNS

The most common way is to create your message payload in a dict and then and to dump it to json in the `publish` call which leads to exactly 0 Message Attributes:
```python
client = boto3.client('sns')
response = client.publish(
    TargetArn=os.getenv('SNS_TO_PUBLISH_TO'),
    Message=json.dumps(my_thing),
    MessageStructure='json'
)
```

Message Attributes are very useful cause you can do filtering on them. But they do lengthen the call needed( or at least require you to create another dict with a specific structure):
```python
client = boto3.client('sns')
response = client.publish(
    TargetArn=os.getenv('SNS_TO_PUBLISH_TO'),
    Message=json.dumps(my_thing),
    MessageStructure='json',
    MessageAttributes={
        'string': {
            'DataType': 'string',
            'StringValue': 'string',
            'BinaryValue': b'bytes'
    }
)
```

We then get message attributes in `event['Records'][0]['Sns']['MessageAttributes']`


### SQS

SQS sending is very similar:
```python
client.send_message(
    QueueUrl='string',
    MessageBody='string',
    MessageAttributes={
        'string': {
            'StringValue': 'string',
            'BinaryValue': b'bytes',
            'StringListValues': [
                'string',
            ],
            'BinaryListValues': [
                b'bytes',
            ],
            'DataType': 'string'
        }
    },
)
```

As you can see from this PR the message attributes are in a different place than on SNS: `event['Records'][0]['messageAttributes']`, of course.

### SNS with SQS

Sometimes you want to send messages to a SNS and then have a couple SQS queues subscribed to that SNS. A fanout pattern.

That comes with its own config options for how Message Attributes are forwarded, where they are forwarded, and if they are forwarded: [this StackOverflow question](https://stackoverflow.com/questions/44238656/how-to-add-sqs-message-attributes-in-sns-subscription) has details on this.

### DynamoDB

Now this is not really messaging. If DynamoDB is used as a DB then it does not really make sense to propagate trace ids.

But some people sometimes use it as a queue too. For example I used it like that for [a demo](https://www.youtube.com/watch?v=-3gOqR_TGEs) where I did want visibility into what went in and not.

DynamoDB can send each change that is done to a DynamoDB stream that invokes a Lambda. That's a really nice way to get kind of SQL-stored-procedures-on-steorids. You can get either just the new row value or you can get the old row and the new row. Or just the keys cause why not.

So you have:
- `event['Records'][0]['dynamodb']['NewImage']`
- `event['Records'][0]['dynamodb']['OldImage']`
- `event['Records'][0]['dynamodb']['keys']`

--- 

And we did not even get into Kinesis, CloudWatch, EventBridge, and other options.

What I am saying is that there are lots of places to search for a context in. Not an infinite list of options so we could code it out, but enough places. That's why I liked giving the user the option to specify a path where they know the context should be.

But I really like the idea of patching the message sending to put them in Message Attributes. On the other hand the Message Attributes are limited to 10 or something so we might get cases where the context won't get propagated.

I am unsure what the best way forward would be. Doing a ton of `json.load()` to get attributes is... not ideal at all. But neither is adding support for a lot of places where context may be.

**Comment by Vlaaaaaaad** [comment]
Aaaaand AWS just added X-Ray( basically tracing) support to SQS by using a Message Attribute too: [Amazon SQS and AWS X-Ray](https://docs.aws.amazon.com/xray/latest/devguide/xray-services-sqs.html).

Still, the scope of the PR seems to be a bit... large. Let's split it up:
- I haven't tested this so I wanna test it next week
    - As soon as that's done I'll report back and get this out of _Draft PR_
    - Maybe adding some docs would be a good idea. Here? Somewhere else?
- This gets merged like it is and requires manual instrumentation of message sending
- A follow-up PR with patching boto lib to inject trace context is created
- With SQS and SNS handled we'll see what the demand is for other events

How does this sound @tredman?

**Comment by tredman** [comment]
Hey - that approach sounds good to me. I can add docs to this page: https://docs.honeycomb.io/getting-data-in/python/beeline/ when done to cover the new functionality.

Sorry for the delay here, wasn't getting notifications for this thread but now I've tasked this so that it's in my queue and I can remember to check back :)

**Comment by Vlaaaaaaad** [comment]
🤦‍♂ now it is me who totally forgot about this. Apologies. ( on a sidenote, this PR discussion is turning into each of us apologising for forgetting stuff 😆 )

I've marked this as a favourite and I should see it in my reports now. I expect I'll finish it during the next two weeks at most.

**Comment by Vlaaaaaaad** [comment]
I went a bit... too deep into testing but progress is being made!

Screen Shot 2019-10-20 at 4 56 53 PM

**Comment by Vlaaaaaaad** [comment]
I am marking this as ready for review as testing was successful. Traces get successfully propagated for most usecases.

**TL;DR**: except for the red things in the second photo this was confirmed working.

We get the following trace:
Screen Shot 2019-10-27 at 6 55 16 PM

for this architecture( the greyed/ red-ish path is not working due to the message being encoded and I could not get rid of the encoding for the life of me):
honeycomb

The demo's pretty disgusting right now, but I'll clean it up and open-source it soon(-ish).

**Comment by tredman** [comment]
LGTM. I can cut a new release now if you want, or do you want to pair it with the boto changes?

**Comment by Vlaaaaaaad** [comment]
> LGTM. I can cut a new release now if you want, or do you want to pair it with the boto changes?

I have pretty much 0 bandwidth to work on that now. I don't think I can get some interns on it either so I'd go for merging this now. I'll create a follow-up PR for automatic boto patching.

**Comment by aidansteele** [comment]
I've been doing something very similar (albeit just for SQS now) in a Golang package: https://github.com/glassechidna/awshoney/blob/605d27da8453012493628d5d5f7b533a4c675d0d/sqs.go#L34

It would be great if we collaborated on ensuring that we use consistent strings for these metadata fields as then we'd have cross-language interop. Thoughts?

**Comment by Vlaaaaaaad** [comment]
@aidansteele sure! That’s why I decided to keep the `X-Honeycomb-Trace` name even though it’s not a HTTP header exactly. 

@tredman how should we sync on this? I remember how the Golang beeline added the `dataset` field to the trace header and the Python beeline was failing to scrape it leading to no context propagation. I’d like to avoid such situations 😅

**Comment by tredman** [comment]
We're working to find an owner for the beeline-interop problem. I don't have an answer yet.

Sigh, I do need to fix that trace propagation issue, though it feels like a breaking change at this point.

**Comment by Vlaaaaaaad** [comment]
For the record for the people that end up here: I did not have time to cleanup the demo code I used to test this. But I open-sourced it at [Vlaaaaaaad/beeline-py-propagation-tests](https://github.com/Vlaaaaaaad/beeline-py-propagation-tests) in hopes at lest it helps someone.
